ci: update codecov workflow to use gcovr

nashif opened the pull request:
- ci: codecov: use gcovr and upload cobertura style file
- kernel: fatal: remove LCOV exclusion


Fixes #66656

nashif committed:
ci: codecov: use gcovr and upload cobertura style file

Use gcovr to generate reports and upload cobertura style file instead of
previously lcov info file.

We have been seeing issues with lcov format and code being reporting as
not covered although by inspecting the generate output files, they
should be.

The XML format we generate is otherwise easier to parse and deal with.

Also reduce frequency to twice a day.

Fixes #66656

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: fatal: remove LCOV exclusion

line is already excluded as part of the function.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

kartben commented:
> ci: codecov: use gcovr and upload cobertura style file

Could you please make the commit message slightly more explicit? Is the change motivated by addressing a particular bug/issue with lcov? Or just moving to a more modern tool? Something else? 
Thanks!

nashif committed:
ci: codecov: add mps2_an385 as a coverage platform

Add one more ARM based platform to get more code coverage.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: add a tag for nano libc tests

tag nano libc tests with nano so we can exclude them if needed. Now they
cause issues when collecting coverage information from tests.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
ci: codecov: increase stack size for tests

Increase stack size for tests and exclude nono libc scenarios causing
some issues with coverage collection.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

Thalley commented:
@nashif Would it make sense to make gcovr the default when we do `twister --coverage`? Seems odd to have CI run the non-default coverage tool

stephanosio commented on .github/workflows/codecov.yaml:
```diff
@@ -1,6 +1,8 @@
 name: Code Coverage with codecov
 
 on:
+  workflow_dispatch:
+    branches: [main]
```
Is this intentional? or a leftover from testing?

nashif replied:
removed.

nashif commented:
> @nashif Would it make sense to make gcovr the default when we do `twister --coverage`? Seems odd to have CI run the non-default coverage tool

yeah, we can do that. @golowanow  can you take care of that?

jhedberg commented on .github/workflows/codecov.yaml:
```diff
@@ -1,6 +1,8 @@
 
 on:
+  workflow_dispatch:
+    branches: [main]
```
It's still here in the first commit. I guess you squashed the removal into the wrong commit?

nashif replied:
I did not squash, thats why I mentioned it in the commit message where it was removed.

nashif replied:
anyways, squashed now.

golowanow commented:
> > @nashif Would it make sense to make gcovr the default when we do `twister --coverage`? Seems odd to have CI run the non-default coverage tool
> 
> yeah, we can do that. @golowanow can you take care of that?

#66712
